src/main/kotlin/frc/robot/subsystems/Drivetrain.kt

```diff
@@ -0,0 +1,198 @@
+import frc.robot.Constants.BLTurnMotorId
+import frc.robot.Constants.BRDriveMotorId
+import frc.robot.Constants.BRTurnEncoderId
+import frc.robot.Constants.BRTurnMotorId
+import frc.robot.Constants.FLTurnEncoderId
+import frc.robot.Constants.FLTurnMotorId
+import frc.robot.Constants.FRTurnEncoderId
+import frc.robot.Constants.FRTurnMotorId
+import frc.robot.commands.DriveCommand
+import frc.robot.controls.ControlScheme
+
+class Drivetrain(
+    val controlScheme: ControlScheme,
+) : SubsystemBase() {
+    init {
+        defaultCommand = DriveCommand(this, controlScheme)
+    }
+
+    private val swerveTab = getTab("Swerve Diagnostics")
+
+    //private PowerDistribution PDP = new PowerDistribution();
+    private val xSpeedEntry = swerveTab.add("xBox xSpeed", 0)
+        .entry
+    private val ySpeedEntry = swerveTab.add("xBox ySpeed", 0)
+        .entry
+    private val rotEntry = swerveTab.add("xBox rot", 0)
+        .entry
+    private val gyroEntry = swerveTab.add("Gyro Heading", 0)
+        .entry
+    val frontLeft = SwerveModule( // front right
+        Constants.FLDriveMotorId,
+        FLTurnMotorId,
+        FLTurnEncoderId,
+        "frontLeft",
+        angleZero = 122.0 + 10.5,
+        position = Translation2d(Constants.MODULE_DISTANCE_X / 2, Constants.MODULE_DISTANCE_Y / 2),
+        controlScheme = controlScheme
+    ) // FIXME: change postion to new drivebase measurements
+    val frontRight = SwerveModule( // backleft
+        Constants.FRDriveMotorId,
+        FRTurnMotorId,
+        FRTurnEncoderId,
+        "frontRight",
+        Translation2d(Constants.MODULE_DISTANCE_X / 2, -Constants.MODULE_DISTANCE_Y / 2),
+        angleZero = 73.0 + 3.1,
+        controlScheme = controlScheme
+
+    ) // FIXME: change postion to new drivebase measurements
+    val backLeft = SwerveModule(
+        BLDriveMotorId,
+        BLTurnMotorId,
+        BLTurnEncoderId,
+        "backLeft",
+        Translation2d(-Constants.MODULE_DISTANCE_X / 2, Constants.MODULE_DISTANCE_Y / 2),
+        angleZero = 65.75 + 4.5,
+        controlScheme = controlScheme
+
+    ) // FIXME: change postion to new drivebase measurements
+    val backRight = SwerveModule(
+        BRDriveMotorId,
+        BRTurnMotorId,
+        BRTurnEncoderId,
+        "backRight",
+        Translation2d(-Constants.MODULE_DISTANCE_X / 2, -Constants.MODULE_DISTANCE_Y / 2),
+        angleZero = 154.5 + 17.25,
+        controlScheme = controlScheme
+
+    ) // FIXME: change postion to new drivebase measurements
+    val modules = listOf(frontLeft, frontRight, backLeft, backRight)
+    val kinematics = SwerveDriveKinematics(
+        *modules.map { it.position }.toTypedArray()
+    )
+
+//    private val gyro: Gyro = ADXRS450_Gyro()
+
+    // Odometry class for tracking robot pose
+    var odometry = SwerveDriveOdometry(
+        kinematics,
+        Rotation2d(),//gyro.rotation2d,
+        modules
+            .map { it.position.toSwerveModulePosition() }
```

Comment:
SwerveModulePosition isn't the position of the module on the robot, but the distance each module's driving motor has traveled. Should be SwerveModulePosition(driveMotor's encoder position in meters, module angle rotation2d)

> The initial positions of the swerve modules (as an array of SwerveModulePosition). In Java, this must be constructed with each wheel position in meters. 

The physical positions (on the robot) of each swerve module are already passed as part of the kinematics object.